### PR TITLE
arch/arm: add support for armclang compiler(AC6) 

### PR DESCRIPTION
**Pull request by anchao**
## Summary

arch/arm: add support for armclang compiler(AC6)

Reference:
https://developer.arm.com/Tools%20and%20Software/Arm%20Compiler%20for%20Embedded

Signed-off-by: chao an <anchao@xiaomi.com>
Signed-off-by: sheyunhu <sheyunhu@xiaomi.com>

## Impact

N/A

## Testing

CI

### CODE REVIEW

**Review by acassis — APPROVED** [review]

File: arch/arm/src/armv7-a/Toolchain.defs [hunk 1]
```diff
@@ -61,4 +61,26 @@ else
   ARCHCPUFLAGS += -mfloat-abi=soft
 endif
 
+ifeq ($(CONFIG_ARM_TOOLCHAIN_ARMCLANG),y)
+
+  ifeq ($(CONFIG_ARCH_CORTEXA5),y)
+    CPUFLAG = Cortex-A5
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-a/Toolchain.defs:67]: CPUFLAG->LDCPUFLAGS?

File: arch/arm/src/armv7-m/Toolchain.defs [hunk 2]
```diff
@@ -75,10 +75,28 @@ ifeq ($(CONFIG_ARM_TOOLCHAIN_CLANG),y)
...
+      endif
+    else
+      LDFLAGS += --cpu=Cortex-M7.no_fp
+    endif
+  else # ifeq ($(CONFIG_ARCH_CORTEXM3),y)
+    LDFLAGS += --cpu=Cortex-M3
+  endif
 
-ifeq ($(CONFIG_ARMV7M_STACKCHECK),y)
-  ARCHOPTIMIZATION += -finstrument-functions -ffixed-r10
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-m/Toolchain.defs]: why remove

File: arch/arm/src/common/Toolchain.defs [hunk 3]
```diff
@@ -174,23 +200,35 @@ endif
 ARCHOPTIMIZATION += -fno-common -Wall -Wshadow -Wundef
 ARCHOPTIMIZATION += -nostdlib
 
-LDFLAGS += -nostdlib
+ifeq ($(CONFIG_ARCH_TOOLCHAIN_CLANG),)
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/Toolchain.defs:203]: CONFIG_ARM_TOOLCHAIN_ARMCLANG

File: boards/misoc/lm32/misoc/README.txt [hunk 4]
```diff
@@ -58,7 +58,7 @@ Buildroot Toolchain
   In order to use the buildroot toolchain, you also must set the following
   in your .config file:
 
-    CONFIG_LM3S_TOOLCHAIN_BUILDROOT=y
+    CONFIG_LM3S_ARM_TOOLCHAIN_BUILDROOT=y
```
[the remark this answers is not in the document]
Comment (anchao) [reply boards/misoc/lm32/misoc/README.txt:61]: Done

File: boards/arm/stm32/stm32f4discovery/scripts/Make.defs [hunk 5]
```diff
@@ -27,7 +27,7 @@ ARCHSCRIPT += $(BOARD_DIR)$(DELIM)scripts$(DELIM)$(LDSCRIPT)
 
 ARCHPICFLAGS = -fpic -msingle-pic-base -mpic-register=r10
 
-ifneq ($(CONFIG_ARM_TOOLCHAIN_CLANG),y)
+ifneq ($(CONFIG_ARCH_TOOLCHAIN_CLANG),y)
```
[the remark this answers is not in the document]
Comment (anchao) [reply boards/arm/stm32/stm32f4discovery/scripts/Make.defs:30]: flag for all clang toolchain, ARMCLANG will select ARCH_TOOLCHAIN_CLANG

File: arch/arm/src/armv7-a/Toolchain.defs (hunk 1 above)
Comment (anchao) [reply arch/arm/src/armv7-a/Toolchain.defs:67]: Done

File: arch/arm/src/armv7-m/Toolchain.defs (hunk 2 above)
Comment (anchao) [reply arch/arm/src/armv7-m/Toolchain.defs]: emm, merge issue, let me restore the change

File: arch/arm/src/common/Toolchain.defs (hunk 3 above)
Comment (anchao) [reply arch/arm/src/common/Toolchain.defs:203]: Done

File: arch/arm/src/Makefile [hunk 6]
```diff
@@ -85,37 +85,49 @@ UOBJS = $(UAOBJS) $(UCOBJS)
...
+  LIBPATH_OPT = -L
 else
-  LDSTARTGROUP ?= --start-group
-  LDENDGROUP   ?= --end-group
+  LIBPATH_OPT = --userlibpath
+  EXTRA_LIBS += arm_vectors.o
 endif
 
-BOARDMAKE = $(if $(wildcard board$(DELIM)Makefile),y,)
+ifeq ($(CONFIG_ARM_TOOLCHAIN_ARMCLANG),y)
```
[the remark this answers is not in the document]
Comment (anchao) [reply arch/arm/src/Makefile]: Done

**Comment by pkarashchenko** [comment]
@anchao do we have any chart comparing code and data size under the equivalent build options between armclang and GCC for example?
I recall we had some chart when adding LLVM

**Comment by anchao** [comment]
> @anchao do we have any chart comparing code and data size under the equivalent build options between armclang and GCC for example? I recall we had some chart when adding LLVM

Yes, the test results will be updated later, let me add the ac6 support for board spresense and imx6 first.

**Pull request merged**